### PR TITLE
Build 3DS, Switch and Wii port with CMake

### PR DESCRIPTION
**Pull request by carstene1ns**

### CODE REVIEW

**Comment by Ghabry** [comment]
I'm pleasantly surprised how small the changes are, somehow I expected this to be much worse (okay, most of the ugliness is probably hidden behind these functions :))

File: CMakeLists.txt [hunk 1]
```diff
@@ -602,7 +643,7 @@ if(PLAYER_BUILD_LIBLCF)
 	target_link_libraries(${PROJECT_NAME} lcf)
 else()
 	# Use system package
-	find_package(liblcf REQUIRED)
+	find_package(liblcf REQUIRED CONFIG)
```
Comment (carstene1ns) [inline CMakeLists.txt:646]: This is actually an unrelated change: We always want a recent liblcf and we always ship a config file, so we can skip all `Findliblcf` stuff.

**Comment by Ghabry** [comment]
Is https://github.com/devkitPro/pacman-packages even maintained anymore? So many open pr, no feedback 🤔

**Comment by Ghabry** [comment]
Well the PR by c1 is "merged"/integrated (kinda) so this can likely progress now?

Will do a build test :)

**Comment by Ghabry** [comment]
Rebased it and added some fixes to make it compile.

After patching devkitpro to use ``BOTH`` instead of ``ONLY`` I get a 3dsx and a nro with the following cmake line:

Switch:
```
cmake . -Bswitch-build -DPLAYER_TARGET_PLATFORM=switch \
  -DCMAKE_PREFIX_PATH="$PWD/../easyrpg-buildscripts/switch" \
  -DCMAKE_TOOLCHAIN_FILE=/opt/devkitpro/cmake/Switch.cmake \
  -DPLAYER_BUILD_LIBLCF=ON \
  -DPLAYER_WITH_FREETYPE=OFF -DPLAYER_WITH_OPUS=OFF
```

There is something broken in the Harfbuzz CMake config file and OPUS has some linker error because of tremor.

3DS:
```
cmake . -B3ds-build -DPLAYER_TARGET_PLATFORM=3ds \
  -DCMAKE_PREFIX_PATH="$PWD/../easyrpg-buildscripts/3ds" \
  -DCMAKE_TOOLCHAIN_FILE=/opt/devkitpro/cmake/3DS.cmake  \
  -DPLAYER_BUILD_LIBLCF=ON \
  -DPLAYER_WITH_FREETYPE=OFF \
  -DPLAYER_WITH_OPUS=OFF
```

**Comment by Ghabry** [comment]
The 3DS build work :partying_face: . Only issue I see is that the consoleInit is wrapped into ``NDEBUG`` so the console is always displayed now. Maybe just use a define in the ui-file that can be commented out to enable this.

For the switch: Here is the NRO, cannot test this one:

[easyrpg-player.nro.zip](https://github.com/EasyRPG/Player/files/8008248/easyrpg-player.nro.zip)

**Comment by carstene1ns** [comment]
Use 'CMAKE_BUILD_TYPE=Release'. ;)

File: CMakeLists.txt [hunk 2]
```diff
@@ -938,6 +939,8 @@ elseif(${PLAYER_TARGET_PLATFORM} MATCHES "^(3ds|switch|psvita)$")
 		ctr_create_3dsx(easyrpg-player
 			SMDH easyrpg-player.smdh)
 	elseif(NINTENDO_SWITCH)
+		target_include_directories(${PROJECT_NAME} PUBLIC
```
[the remark this answers is not in the document]
Comment (Ghabry) [reply CMakeLists.txt:949]: The toolchain only sets ``-L`` but not ``-I``.
The proper way would be to request all the libraries through pkg-config (as we can be certain pkg-config is available due to devkitpro). That was the first thing I tried but CMake dislikes the pc files because some of them contain ``prefix=$(devkitpro)...`` instead of ``prefix=/opt/devkitpro``.

So would require asking DKP maintainers to patch all the pc files :/

File: CMakeLists.txt [hunk 3]
```diff
@@ -986,12 +989,14 @@ elseif(${PLAYER_TARGET_PLATFORM} MATCHES "^(3ds|switch|psvita)$")
 			FILE "${CMAKE_CURRENT_SOURCE_DIR}/resources/psvita/touch_ui.png" touch_ui.png)
 	endif()
 	# for distribution
-	add_custom_command(TARGET easyrpg-player POST_BUILD
-		BYPRODUCTS easyrpg-player.elf
-		COMMAND ${CMAKE_COMMAND} -E copy_if_different
-			easyrpg-player easyrpg-player.elf
-		WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
-		COMMENT "Prepare for distribution")
+	if(EXISTS "${CMAKE_CURRENT_BINARY_DIR}/easyrpg-player")
```
[the remark this answers is not in the document]
Comment (Ghabry) [reply CMakeLists.txt:999]: When using DKP toolchain the file is resulting file is already called "easyrpg-player.elf" so I'm not sure if we even need this custom command anymore. Can also remove it.

[the remark this answers is not in the document]
Comment (carstene1ns) [reply CMakeLists.txt:999]: That would be okay. I added it just because we have shipped the elf ever since and it would otherwise require a `cp` on jenkins. But since it is only used now by vita, we could also likely set `CMAKE_EXECUTABLE_SUFFIX=.elf` there and remove this block.

[the remark this answers is not in the document]
Comment (carstene1ns) [reply CMakeLists.txt:999]: You only get elf when you have recent cmake, btw: https://github.com/devkitPro/pacman-packages/blob/7a3febdaa5e784bce61bbbdcad94d9eb65e828f2/cmake/common-utils/dkp-toolchain-common.cmake#L9-L11

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ghabry) [reply CMakeLists.txt:949]: Actually there is a CMake config file for use with ``find_package(OpenGL CONFIG REQUIRED`` but the config file is not found because it is not in one of the search directories used by CMake.

I'm not sure if this is a CMake bug or not.

I asked for clarification here: https://gitlab.kitware.com/cmake/cmake/-/issues/23192

[the remark this answers is not in the document]
Comment (Ghabry) [reply CMakeLists.txt:949]: The ``find_package`` problem was resolved in the recent dkp cmake update. Solving the last blocker.

**Comment by Ghabry** [comment]
They also added a CMake Toolchain for Wii now.

@carstene1ns could you update dkp on your buildbot to the latest version?

Looks all pretty good now to give this the final push and then we can go all cmake :)

File: src/platform/3ds/3ds_ui.cpp [hunk 4]
```diff
@@ -126,9 +123,9 @@ CtrUi::CtrUi(int width, int height, const Game_ConfigVideo& cfg) : BaseUi(cfg)
 #endif
 
 #ifdef NDEBUG
-	// Loading bottom screen keyboard
-	assets = C2D_SpriteSheetLoadFromMem(assets_t3x, assets_t3x_size);
-	bottom_image = C2D_SpriteSheetGetImage(assets, assets_keyboard_idx);
+	// FIXME: Loading bottom screen keyboard, no spritesheet!
```
Comment (carstene1ns) [inline src/platform/3ds/3ds_ui.cpp:126]: This was actually a hack to begin with. Using a spritesheet with only one image to avoid setting up all texture and image structs. `ctr_add_graphics_target` does not like this (and while we could work around it), doing it manually is likely better. Will come up with a solution for console vs. keyboard anyway, since cmake does not add `NDEBUG`, likely because of some custom flags.

**Review by carstene1ns — COMMENTED** [review]
Btw. Wii is untested, need to setup the toolchain locally tonight.

**Comment by carstene1ns** [comment]
my make_wii.sh:
```bash
#!/bin/sh

set -e

toolchaindir="/home/carsten/projects/easyrpg/buildscripts/wii"
lcfdir="/home/carsten/projects/easyrpg/liblcf/wii"

#export PKG_CONFIG="pkg-config --static"

if [ "x$1" == "xclean" ]; then
  rm -rf build-wii
  exit
fi

cmake -B build-wii . -G Ninja -DCMAKE_TOOLCHAIN_FILE=$DEVKITPRO/cmake/Wii.cmake \
  -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_BUILD_TYPE=Release -DPLAYER_TARGET_PLATFORM=wii \
  -DCMAKE_PREFIX_PATH="$toolchaindir;$lcfdir" -DPLAYER_WITH_FREETYPE=NO
cmake --build build-wii
```

Currently the dol is 12 mb, which seems a bit too much. Need to investigate.

File: builds/cmake/Modules/PlayerBuildType.cmake [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	endif()
+	message(STATUS "Setting build type to '${default_build_type}' as none was specified.")
+	set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
+	set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
+else()
+	message(STATUS "Build type is set to ${CMAKE_BUILD_TYPE}.")
+endif()
+
+# Ensure needed debugging flags are passed
+add_compile_definitions($<$<CONFIG:Debug>:_DEBUG>$<$<CONFIG:Release,MinSizeRel,RelWithDebInfo>:NDEBUG>)
```
Comment (carstene1ns) [inline builds/cmake/Modules/PlayerBuildType.cmake:16]: Available since cmake 3.12 :/
Well, easy fix.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (carstene1ns) [reply CMakeLists.txt]: Interestingly enough it works locally, but not on my ports jenkins agent.

**Comment by carstene1ns** [comment]
One thing currently missing is romfs support, but this can be added later.

**Comment by Ghabry** [comment]
@carstene1ns 
Works great except for one minor thing:
The Switch section needs the following change, otherwise the ``EGL.h`` header is not found in ``switch_ui.cpp``

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 109d066c..c3ac39b3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -532,6 +532,7 @@ elseif(${PLAYER_TARGET_PLATFORM} STREQUAL "switch")
 		src/platform/switch/switch_ui.cpp
 		src/platform/switch/switch_ui.h
 		${PLAYER_SWITCH_ASSETS})
+	target_link_libraries(${PROJECT_NAME} OpenGL::EGL)
 elseif(${PLAYER_TARGET_PLATFORM} STREQUAL "wii")
 	if(NOT NINTENDO_WII)
 		message(FATAL_ERROR "Missing toolchain file! Use '-DCMAKE_TOOLCHAIN_FILE=$DEVKITPRO/cmake/Wii.cmake' option.")
@@ -955,8 +956,7 @@ elseif(${PLAYER_TARGET_PLATFORM} MATCHES "^(psvita|3ds|switch|wii)$")
 	elseif(NINTENDO_SWITCH)
 		target_link_libraries(easyrpg-player
 			${PROJECT_NAME}
-			${GLAD}
-			OpenGL::EGL)
+			${GLAD})
 		nx_generate_nacp(easyrpg-player.nacp
 			NAME "EasyRPG Player"
 			AUTHOR "EasyRPG Team"
```

**Review by fdelapena — APPROVED** [review]

**Review by Ghabry — APPROVED** [review]
